File: summer2019/week1/valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int lo = 0;
        int hi = s.length() - 1;
        
        while (lo < hi) {
            if (!isAlphanumeric(s.charAt(lo))) {
                lo++;
                continue;
            }
            if (!isAlphanumeric(s.charAt(hi))) {
                hi--;
                continue;
            }
            if (Character.toLowerCase(s.charAt(lo)) != Character.toLowerCase(s.charAt(hi))) {
                System.out.println(s.charAt(lo));
                System.out.println(s.charAt(hi));
                return false;
            }
            hi--;
            lo++;
        }
        return true;
    }

    public boolean isAlphanumeric(char c) {
        return Character.isDigit(c) || Character.isLetter(c);
    }
}

File: spring2020/unit1/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        lo = 0
        hi = len(s) - 1
        while lo <= hi:
            if not s[lo].isalnum():
                lo += 1
                continue
            if not s[hi].isalnum():
                hi -= 1
                continue
            if s[lo].lower() != s[hi].lower():
                return False
            lo += 1
            hi -= 1
        return True
